core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -89,19 +88,35 @@ class TransactionStateManager(brokerId: Int,
   /**
    * Get the transaction metadata associated with the given transactional id, or null if not found
    */
-  def getTransactionState(transactionalId: String): Option[TransactionMetadata] = {
-    Option(transactionMetadataCache.get(transactionalId))
+  def getTransactionState(transactionalId: String): Option[CoordinatorEpochAndTxnMetadata] = {
+    val partitionId = partitionFor(transactionalId)
+
+    transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
```

Comment:
1. Please see my previous comment: we will modify the `statelock` into a read-write lock and the only operations that could mutate the top-level map `transactionMetadataCache` will be covered (the checking / append-to-log will be covered in read lock).

2. For the lower-level map `Pool[String, TransactionMetadata]`, its operations are thread safe and any modifications to the inner `TransactionMetadata` will be covered by the object `synchronized` itself and all modifications will be in-place than object override.